File: static/js/ninja-trader.js
// NinjaTrader API Integration

class NinjaTraderAPI {
    constructor() {
        this.statusElement = document.getElementById('nt-status');
        // Wrap initialization in try-catch to handle any initialization errors
        try {
            this.initialize();
        } catch (error) {
            console.warn('Error initializing NinjaTrader API:', error);
            this.updateStatus('Error initializing NinjaTrader connection', 'error');
        }
    }

    initialize() {
        this.checkStatus();
        // Check status every 30 seconds
        setInterval(() => {
            try {
                this.checkStatus();
            } catch (error) {
                console.warn('Error checking NinjaTrader status:', error);
            }
        }, 30000);
    }

    async checkStatus() {
        if (!this.statusElement) {
            console.warn('Status element not found');
            return false;
        }

        try {
            const response = await fetch('/api/ninja-trader/status');
            const data = await response.json();
            
            this.updateStatus(
                data.available ? data.message : 'NinjaTrader integration unavailable',
                this.getStatusClass(data)
            );
            
            if (data.error) {
                this.statusElement.title = data.error;
            }
            
            return data.connected;
        } catch (error) {
            console.warn('Error checking NinjaTrader status:', error);
            this.updateStatus('Error connecting to NinjaTrader', 'error');
            return false;
        }
    }

    updateStatus(message, statusClass) {
        if (this.statusElement) {
            this.statusElement.textContent = message;
            this.statusElement.className = `nt-status status-${statusClass}`;
        }
    }

    getStatusClass(data) {
        if (!data.available) return 'unavailable';
        if (data.connected) return 'connected';
        return 'disconnected';
    }

    async getRecentTrades(days = 7, account = null) {
        try {
            const params = new URLSearchParams({
                days: days.toString()
            });
            if (account) {
                params.append('account', account);
            }

            const response = await fetch(`/api/ninja-trader/trades?${params}`);
            const data = await response.json();
            
            if (!data.success) {
                throw new Error(data.error || 'Unknown error fetching trades');
            }
            
            return data.trades;
        } catch (error) {
            console.warn('Error getting recent trades:', error);
            throw error;
        }
    }

    async getMarketData(instrument, timeframe = '1 Day', days = 30) {
        try {
            const params = new URLSearchParams({
                instrument,
                timeframe,
                days: days.toString()
            });

            const response = await fetch(`/api/ninja-trader/market-data?${params}`);
            const data = await response.json();
            
            if (!data.success) {
                throw new Error(data.error || 'Unknown error fetching market data');
            }
            
            return data.data;
        } catch (error) {
            console.warn('Error getting market data:', error);
            throw error;
        }
    }
}

// Initialize NinjaTrader API integration when DOM is loaded
document.addEventListener('DOMContentLoaded', () => {
    try {
        window.ntAPI = new NinjaTraderAPI();
    } catch (error) {
        console.warn('Error creating NinjaTrader API instance:', error);
    }
});

File: routes/ninja_trader.py
from flask import Blueprint, jsonify, request

# Initialize variables
NINJA_TRADER_AVAILABLE = False
NINJA_TRADER_ERROR = None
NinjaTraderAPI = None

# Try to import NinjaTrader APIs
try:
    from ninja_trader_api import NinjaTraderAPI
    from ninjatrader_dll_api import NinjaTraderDLL
    NINJA_TRADER_AVAILABLE = True
    USE_DLL = True  # Set to True to use DLL API, False to use socket API
except ImportError as e:
    NINJA_TRADER_ERROR = str(e)
    print(f"Warning: NinjaTrader API not available: {e}")
except Exception as e:
    NINJA_TRADER_ERROR = str(e)
    print(f"Error initializing NinjaTrader API: {e}")
from ExecutionProcessing import process_trades
import json
from datetime import datetime, timedelta
from config import config

ninja_trader_bp = Blueprint('ninja_trader', __name__)

# Cache for NinjaTrader API instance
_nt_api = None

def get_nt_api():
    """Get or create NinjaTrader API instance."""
    global _nt_api
    if _nt_api is None:
        try:
            if USE_DLL:
                print("Using NinjaTrader DLL API")
                _nt_api = NinjaTraderDLL()
            else:
                print("Using NinjaTrader Socket API")
                _nt_api = NinjaTraderAPI()
        except Exception as e:
            print(f"Error creating NinjaTrader API: {e}")
            return None
    return _nt_api

@ninja_trader_bp.route('/api/ninja-trader/status', methods=['GET'])
def get_status():
    """Check NinjaTrader connection status."""
    if not NINJA_TRADER_AVAILABLE:
        return jsonify({
            'connected': False,
            'available': False,
            'error': NINJA_TRADER_ERROR or 'NinjaTrader API not available',
            'message': 'NinjaTrader integration is not available'
        })

    api = get_nt_api()
    return jsonify({
        'connected': bool(api and api.connected),
        'available': True,
        'error': None if api and api.connected else 'Not connected to NinjaTrader',
        'message': 'Connected to NinjaTrader' if api and api.connected else 'Not connected to NinjaTrader'
    })

@ninja_trader_bp.route('/api/ninja-trader/trades', methods=['GET'])
def get_trades():
    """Get recent trades from NinjaTrader."""
    api = get_nt_api()
    if not api or not api.connected:
        return jsonify({
            'success': False,
            'error': 'NinjaTrader API not available'
        }), 503

    try:
        # Get parameters
        days = request.args.get('days', '7', type=int)
        account = request.args.get('account', None)
        start_date = datetime.now() - timedelta(days=days)

        # Get execution data
        df = api.get_executions(start_date=start_date, account=account)
        
        # Load instrument multipliers
        with open(config.instrument_config, 'r') as f:
            multipliers = json.load(f)

        # Process trades
        processed_trades = process_trades(df, multipliers)

        return jsonify({
            'success': True,
            'trades': processed_trades
        })

    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

@ninja_trader_bp.route('/api/ninja-trader/market-data', methods=['GET'])
def get_market_data():
    """Get market data for a specific instrument."""
    api = get_nt_api()
    if not api or not api.connected:
        return jsonify({
            'success': False,
            'error': 'NinjaTrader API not available'
        }), 503

    try:
        # Get parameters
        instrument = request.args.get('instrument', type=str)
        timeframe = request.args.get('timeframe', '1 Day', type=str)
        days = request.args.get('days', '30', type=int)
        
        if not instrument:
            return jsonify({
                'success': False,
                'error': 'Instrument parameter is required'
            }), 400

        start_date = datetime.now() - timedelta(days=days)
        
        # Get market data
        bars = api.get_bars(
            instrument=instrument,
            start_date=start_date,
            timeframe=timeframe
        )

        return jsonify({
            'success': True,
            'data': bars
        })

    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

@ninja_trader_bp.errorhandler(Exception)
def handle_error(error):
    """Global error handler for ninja_trader blueprint."""
    return jsonify({
        'success': False,
        'error': str(error)
    }), 500